File: basic-tut-8.py
#!/usr/bin/env python3
import sys
import logging
import gi
import signal

gi.require_version("GLib", "2.0")
gi.require_version("GObject", "2.0")
gi.require_version("Gst", "1.0")
gi.require_version("GstBase", "1.0")
gi.require_version("GstAudio", "1.0")
from gi.repository import GLib, GObject, Gst, GstAudio

logging.basicConfig(
    level=logging.DEBUG, format="[%(name)s] [%(levelname)8s] - %(message)s"
)
logger = logging.getLogger(__name__)


class CustomData:
    def __init__(self):
        self.app_source = Gst.ElementFactory.make("appsrc", "audio_source")
        self.tee = Gst.ElementFactory.make("tee", "tee")
        self.audio_queue = Gst.ElementFactory.make("queue", "audio_queue")
        self.audio_convert1 = Gst.ElementFactory.make("audioconvert", "audio_convert1")
        self.audio_resample = Gst.ElementFactory.make("audioresample", "audio_resample")
        self.audio_sink = Gst.ElementFactory.make("autoaudiosink", "audio_sink")
        self.video_queue = Gst.ElementFactory.make("queue", "video_queue")
        self.audio_convert2 = Gst.ElementFactory.make("audioconvert", "audio_convert2")
        self.visual = Gst.ElementFactory.make("wavescope", "visual")
        self.video_convert = Gst.ElementFactory.make("videoconvert", "video_convert")
        self.video_sink = Gst.ElementFactory.make("autovideosink", "video_sink")
        self.app_queue = Gst.ElementFactory.make("queue", "app_queue")
        self.app_sink = Gst.ElementFactory.make("appsink", "app_sink")
        self.pipeline = Gst.Pipeline.new("test-pipeline")
        self.a, self.b, self.c, self.d = (
            0,
            0,
            0,
            0,
        )  # for waveform generation
        self.sourceid = 0  # to control GSource
        self.num_samples = (
            0  # number of samples generated so far (for time stamp generation)
        )
        self.main_loop = None


def push_data(data):
    chunk_size = 1024
    num_samples = chunk_size // 2  # because each sample is 16 bits
    # Create a new empty buffer
    buffer = Gst.Buffer.new_allocate(None, chunk_size, None)

    # To-Do: set buffer timestamp and duration
    
    # Generate some psychedelic waveforms aka make buffer data
    ret, map_info = buffer.map(Gst.MapFlags.WRITE)
    raw = map_info.data
    #print("Raw:", raw)
    print("Raw Type", type(raw))
    raw = list(raw)
    #print("Raw:", raw)
    print("Raw Type", type(raw))
    data.c += data.d
    data.d -= data.c // 1000
    freq = 1100 + 1000 * data.d
    for i in range(num_samples):
        data.a += data.b
        data.b -= data.a // freq
        raw[i] = int(500 * data.a)

    data.num_samples += num_samples

    # Push the buffer into the appsrc
    ret = data.app_source.emit("push-buffer", buffer)
    if ret != Gst.FlowReturn.OK:
        return False

    return True


def start_feed(object, arg0, data):
    if data.sourceid == 0:
        print("Start feeding\n")
        # data.sourceid =
        GLib.timeout_add(10, push_data, data)
        data.sourceid = 1


def stop_feed(object, data):
    if data.sourceid != 0:
        print("Stop feeding\n")
        GObject.source_remove(data.sourceid)
        data.sourceid = 0


def new_sample(object, data):
    sample = data.app_sink.emit("pull-sample")
    if sample:
        print("*")
        return Gst.FlowReturn.OK

    return Gst.FlowReturn.ERROR


def error_cb(bus, msg, data):
    err, debug_info = msg.parse_error()
    logger.error(f"Error received from element {msg.src.get_name()}: {err.message}")
    logger.error(f"Debugging information: {debug_info if debug_info else 'none'}")
    data.main_loop.quit()


def main():
    Gst.init(None)
    data = CustomData()
    if (
        not data.pipeline
        or not data.app_source
        or not data.tee
        or not data.audio_queue
        or not data.audio_convert1
        or not data.audio_resample
        or not data.audio_sink
        or not data.video_queue
        or not data.audio_convert2
        or not data.visual
        or not data.video_convert
        or not data.video_sink
        or not data.app_queue
        or not data.app_sink
    ):
        logger.error("Not all elements could be created")
        sys.exit(1)

    # Configure wavescope
    data.visual.set_property("shader", 0)
    data.visual.set_property("style", 0)

    # Configure appsrc
    sample_rate = 44100
    info = GstAudio.AudioInfo()
    info.set_format(GstAudio.AudioFormat.S16, sample_rate, 1, None)
    audio_caps = info.to_caps()
    data.app_source.set_property("caps", audio_caps)
    data.app_source.set_property("format", Gst.Format.TIME)
    data.app_source.connect("need-data", start_feed, data)
    data.app_source.connect("enough-data", stop_feed, data)

    # configure appsink
    data.app_sink.set_property("emit-signals", True)
    data.app_sink.set_property("caps", audio_caps)
    data.app_sink.connect("new-sample", new_sample, data)

    # add elements to pipeline
    data.pipeline.add(
        data.app_source,
        data.tee,
        data.audio_queue,
        data.audio_convert1,
        data.audio_resample,
        data.audio_sink,
        data.video_queue,
        data.audio_convert2,
        data.visual,
        data.video_convert,
        data.video_sink,
        data.app_queue,
        data.app_sink,
    )
    # Link elements with "Always" pads
    ret1 = data.app_source.link(data.tee)
    ret2 = (
        data.audio_queue.link(data.audio_convert1)
        and data.audio_convert1.link(data.audio_resample)
        and data.audio_resample.link(data.audio_sink)
    )
    ret3 = (
        data.video_queue.link(data.audio_convert2)
        and data.audio_convert2.link(data.visual)
        and data.visual.link(data.video_convert)
        and data.video_convert.link(data.video_sink)
    )
    ret4 = data.app_queue.link(data.app_sink)

    if not ret1 or not ret2 or not ret3 or not ret4:
        logger.error("Elements could not be linked")
        sys.exit(1)

    # Manually link tee, which has request pads
    tee_audio_pad = data.tee.get_request_pad("src_%u")
    logger.info("Obtained request pad for audio branch")  # To do: add gst_pad_get_name
    queue_audio_pad = data.audio_queue.get_static_pad("sink")

    tee_video_pad = data.tee.get_request_pad("src_%u")
    logger.info("Obtained request pad for video branch")  # To do: add gst_pad_get_name
    queue_video_pad = data.video_queue.get_static_pad("sink")

    tee_app_pad = data.tee.get_request_pad("src_%u")
    logger.info("Obtained request pad for app branch")
    queue_app_pad = data.app_queue.get_static_pad("sink")

    ret_audio = tee_audio_pad.link(queue_audio_pad)
    ret_video = tee_video_pad.link(queue_video_pad)
    ret_app = tee_app_pad.link(queue_app_pad)
    if ret_audio != 0 or ret_video != 0 or ret_app != 0:
        logger.error("Tee could not be linked")
        sys.exit(1)

    # Setup bus and message handlers
    bus = data.pipeline.get_bus()
    bus.add_signal_watch()
    bus.connect("message::error", error_cb, data)

    # Play pipeline
    data.pipeline.set_state(Gst.State.PLAYING)
    print("Pipeline play")

    # Create a GLib Main loop and set it to return
    data.main_loop = GLib.MainLoop()
    data.main_loop.run()


if __name__ == "__main__":
    main()

File: C-code/basic-tut8.c
#include<gst/gst.h>
#include<gst/audio/audio.h>
#include <string.h>

#define CHUNK_SIZE 1024
#define SAMPLE_RATE 44100

typedef struct _CustomData{
  GstElement *app_source;
  GstElement *tee;
  GstElement *audio_queue;
  GstElement *audio_convert1;
  GstElement *audio_resample;
  GstElement *audio_sink;
  GstElement *video_queue;
  GstElement *audio_convert2;
  GstElement *visual;
  GstElement *video_convert;
  GstElement *video_sink;
  GstElement *app_queue;
  GstElement *app_sink;
  GstElement *pipeline;
  GMainLoop *main_loop;
  guint64 num_samples;
  gfloat a, b, c, d;
  guint sourceid;
} CustomData;

// Super-basic square wave
static gboolean pushdata(CustomData *data){
  GstBuffer *buffer;
  GstFlowReturn ret;
  GstMapInfo map;
  gint16 *raw;
  gint num_samples = CHUNK_SIZE / 2; // Each sample is 16 bits = 2 bytes
  gfloat freq;

  buffer = gst_buffer_new_and_alloc(CHUNK_SIZE);
  //Timestamp and duration for buffer
  GST_BUFFER_TIMESTAMP(buffer) = gst_util_uint64_scale(data->num_samples, GST_SECOND, SAMPLE_RATE);
  GST_BUFFER_DURATION(buffer) = gst_util_uint64_scale(num_samples, GST_SECOND, SAMPLE_RATE);

  // Generating waveforms
  gst_buffer_map (buffer, &map, GST_MAP_WRITE);
  raw = (gint16*)map.data;
  data->c += data->d;
  data->d -= data->c / 1000;
  freq = 1100 + 1000 * data->d;
  for(int i = 0; i < num_samples; i++){
    data->a += data->b;
    data->b -= data->a / freq;
    raw[i] = (gint16)(data->b);
  }
  gst_buffer_unmap (buffer, &map);
  data->num_samples += num_samples * 2;

  g_signal_emit_by_name(data->app_source, "push-buffer", buffer, &ret);
  gst_buffer_unref(buffer);

  if (ret != GST_FLOW_OK){
    return FALSE;
  }
  return TRUE;
}
// Sine wave
/*static gboolean pushdata(CustomData *data){
  GstBuffer *buffer;
  GstFlowReturn ret;
  GstMapInfo map;
  gint16 *raw;
  gint num_samples = CHUNK_SIZE / 2; // Each sample is 16 bits = 2 bytes
  gfloat freq;

  buffer = gst_buffer_new_and_alloc(CHUNK_SIZE);
  //Timestamp and duration for buffer
  GST_BUFFER_TIMESTAMP(buffer) = gst_util_uint64_scale(data->num_samples, GST_SECOND, SAMPLE_RATE);
  GST_BUFFER_DURATION(buffer) = gst_util_uint64_scale(num_samples, GST_SECOND, SAMPLE_RATE);

  // Generating waveforms
  gst_buffer_map (buffer, &map, GST_MAP_WRITE);
  raw = (gint16*)map.data;
  data->c += data->d;
  data->d -= data->c / 1000;
  freq = 1100 + 1000 * data->d;
  for(int i = 0; i < num_samples; i++){
    data->a += data->b;
    data->b -= data->a / freq;
    raw[i] = (gint16)(500 * data->a);
  }
  gst_buffer_unmap (buffer, &map);
  data->num_samples += num_samples;

  g_signal_emit_by_name(data->app_source, "push-buffer", buffer, &ret);
  gst_buffer_unref(buffer);

  if (ret != GST_FLOW_OK){
    return FALSE;
  }
  return TRUE;
}*/

static void start_feed(GstElement *source, guint size, CustomData *data){
  if(data->sourceid == 0){
    g_print("Start data feed\n");
    data->sourceid = g_idle_add((GSourceFunc)pushdata, data); //g_idle_add returns a guint? The ID of the event source
  }
}

static void stop_feed(GstElement *source, CustomData *data){
  if(data->sourceid != 0){
    g_print("Stop data feed\n");
    g_source_remove(data->sourceid); //g_source_remove removed the source with given ID from the default main context
    data->sourceid = 0;
  }
}

static GstFlowReturn new_sample(GstElement *sink, CustomData *data){
  GstSample *sample;
  g_signal_emit_by_name(sink, "pull-sample", &sample);
  if(sample){
    g_print("*");
    gst_sample_unref(sample);
    return GST_FLOW_OK;
  }
  return GST_FLOW_ERROR;
}
static void error_cb(GstBus *bus, GstMessage *msg, CustomData *data){
  GError *err;
  gchar *debug_info;

  gst_message_parse_error(msg, &err, &debug_info);
  g_printerr ("Error received from element %s: %s\n", GST_OBJECT_NAME (msg->src), err->message);
  g_printerr ("Debugging information: %s\n", debug_info ? debug_info : "none");
  g_clear_error (&err);
  g_free (debug_info);

  g_main_loop_quit (data->main_loop);
}


int main(int argc, char *argv[]){
  gst_init(&argc, &argv);
  CustomData data;
  GstAudioInfo info;
  memset(&data, 0, sizeof(data));
  data.b = 1;
  data.d = 1;
  data.app_source = gst_element_factory_make("appsrc", "audio_source");
  data.tee = gst_element_factory_make("tee", "tee");
  data.audio_queue = gst_element_factory_make("queue", "audio_queue");
  data.audio_convert1 = gst_element_factory_make("audioconvert", "audio_convert1");
  data.audio_resample = gst_element_factory_make("audioresample", "audio_resample");
  data.audio_sink = gst_element_factory_make("autoaudiosink", "audio_sink");
  data.video_queue = gst_element_factory_make("queue", "video_queue");
  data.audio_convert2 = gst_element_factory_make("audioconvert", "audio_convert2");
  data.visual = gst_element_factory_make("wavescope", "visual");
  data.video_convert = gst_element_factory_make("videoconvert", "video_convert");
  data.video_sink = gst_element_factory_make("autovideosink", "video_sink");
  data.app_queue = gst_element_factory_make("queue", "app_queue");
  data.app_sink = gst_element_factory_make("appsink", "app_sink");
  data.pipeline = gst_pipeline_new("test-pipeline");

if(!data.app_source || !data.tee || !data.audio_queue || !data.audio_convert1 || !data.audio_resample || !data.audio_sink || !data.video_queue || !data.audio_convert2 || !data.visual || !data.video_convert || !data.video_sink || !data.app_queue || !data.app_sink || !data.pipeline){
  g_printerr("All Elements could not be created \n");
  return -1;
}

  // configure wavescope visual
  g_object_set (data.visual, "shader", 0, "style", 0, NULL);

 //configure app_source caps
 gst_audio_info_set_format(&info, GST_AUDIO_FORMAT_S16, SAMPLE_RATE, 1, NULL);
GstCaps *audio_caps = gst_audio_info_to_caps(&info);
g_object_set(data.app_source, "caps", audio_caps, "format", GST_FORMAT_TIME, NULL);
// configure app_source signal handler
g_signal_connect(data.app_source, "need-data", G_CALLBACK(start_feed), &data);
g_signal_connect(data.app_source, "enough-data", G_CALLBACK(stop_feed), &data);

 //configure app_sink
 //emit-signals needs to be set to true first
 g_object_set(data.app_sink, "emit-signals", TRUE, "caps", audio_caps, NULL);
 //configure app_sink signal handler
 g_signal_connect(data.app_sink, "new-sample", G_CALLBACK(new_sample), &data);
 gst_caps_unref(audio_caps);

// Link elements with Always pads
gst_bin_add_many(GST_BIN(data.pipeline), data.app_source, data.tee, data.audio_queue, data.audio_convert1, data.audio_resample, data.audio_sink, data.video_queue, data.audio_convert2, data.visual, data.video_convert, data.video_sink, data.app_queue, data.app_sink, NULL);
gboolean link1 = gst_element_link_many(data.app_source, data.tee, NULL);
gboolean link2 = gst_element_link_many(data.audio_queue, data.audio_convert1, data.audio_resample, data.audio_sink, NULL);
gboolean link3 = gst_element_link_many(data.video_queue, data.audio_convert2, data.visual, data.video_convert, data.video_sink, NULL);
gboolean link4 = gst_element_link_many(data.app_queue, data.app_sink, NULL);

if(!link1 || !link2 || !link3 || !link4){
  g_printerr("Elements could not be linked \n");
  return -1;
}

//Manual linking with tee pads
GstPad *tee_audio_pad;
tee_audio_pad = gst_element_get_request_pad(data.tee, "src_%u");
GstPad *queue_audio_pad = gst_element_get_static_pad(data.audio_queue, "sink");

GstPad *tee_video_pad = gst_element_get_request_pad(data.tee, "src_%u");
GstPad *queue_video_pad = gst_element_get_static_pad(data.video_queue, "sink");

GstPad *tee_app_pad = gst_element_get_request_pad(data.tee, "src_%u");
GstPad *queue_app_pad = gst_element_get_static_pad(data.app_queue, "sink");
gst_pad_link(tee_audio_pad, queue_audio_pad);
gst_pad_link(tee_video_pad, queue_video_pad);
gst_pad_link(tee_app_pad, queue_app_pad);

gst_object_unref(queue_audio_pad);
gst_object_unref(queue_video_pad);
gst_object_unref(queue_app_pad);


//Connect bus to error-handler callback
//Signals to notify application of error events
GstBus *bus = gst_element_get_bus(data.pipeline);
gst_bus_add_signal_watch(bus);
g_signal_connect(G_OBJECT(bus), "message::error", (GCallback)error_cb, &data);
gst_object_unref(bus);

//Start playing pipeline
gst_element_set_state(data.pipeline, GST_STATE_PLAYING);

//GLib main loop
data.main_loop = g_main_loop_new(NULL, FALSE);
g_main_loop_run(data.main_loop);


// Freeing resources
gst_element_release_request_pad (data.tee, tee_audio_pad);
gst_element_release_request_pad (data.tee, tee_video_pad);
gst_element_release_request_pad (data.tee, tee_app_pad);
gst_object_unref (tee_audio_pad);
gst_object_unref (tee_video_pad);
gst_object_unref (tee_app_pad);
gst_element_set_state(data.pipeline, GST_STATE_NULL);
gst_object_unref(data.pipeline);

return 0;
}
